### PR TITLE
Add bouncycastle permissions to security.policy

### PR DESCRIPTION
**Pull request by cwperks**
### Description

Adds bouncycastle permissions into core's `security.policy` file. With the introduction of opensearch-encryption-sdk, the bouncycastle provider (`bcprovp-jdk15to18`) is now a dependency of core. The security plugin also has a direct dependency on the same artifact and has permissions in its `plugin-security.policy` file to grant the security plugin permission to add the bouncy castle provider.

There is an issue bringing up a node with the security plugin installed and TLS enabled because the permissions are not present in `bootstrap/security.policy` and the dependency has now been moved to core

### Related Issues
Resolves https://github.com/opensearch-project/security/issues/3309

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by cwperks** [comment]
@peternied @reta @Bukhtawar @gbbafna @vikasvb90

**Comment by cwperks** [comment]
Tested by performing plugin install manually of the security plugin.

- Make a local distribution of core: `./gradlew localDistro`
- `cd build/distribution/local/opensearch-3.0.0-SNAPSHOT`
- Make a snapshot of the security plugin. From the root directory of the security repo run `./gradlew assemble`
- Plugin output is at `<OPENSEARCH_SECURITY_HOME>/build/distributions/opensearch-security-3.0.0.0-SNAPSHOT.zip`
- Install the plugin. Example: `./bin/opensearch-plugin install file:/Users/cwperx/Projects/opensearch/security/build/distributions/opensearch-security-3.0.0.0-SNAPSHOT.zip`
- Run the `install_demo_configuration.sh` script and answer 'y' to all questions: `./plugins/opensearch-security/tools/install_demo_configuration.sh`
- Run OpenSearch node: `./bin/opensearch`

There should be no errors on startup. The security plugin has an automated CI check for plugin install, but the check will not work until this PR is merged.

**Review by peternied — COMMENTED** [review]
Is there a danger that this lowers the security posture of the SSL in such a way that **any** plugins can manipulate the SSL configuration, whereas previously this was isolated to plugins with this policy?

**Comment by cwperks** [comment]
> Is there a danger that this lowers the security posture of the SSL in such a way that any plugins can manipulate the SSL configuration, whereas previously this was isolated to plugins with this policy?

Yes that's correct. Any plugin would be able to execute this same block as the security plugin: https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/OpenSearchSecurityPlugin.java#L334-L342

With the `plugin-security.policy` file, cluster admins installing plugins will accept a policy when installing a plugin. That is not the case with the policy file in this PR. The `security.policy` file in this PR is applicable to core and plugins.

File: server/src/main/resources/org/opensearch/bootstrap/security.policy
Comment (RyanL1997) [inline server/src/main/resources/org/opensearch/bootstrap/security.policy:88]: Do we need this for jdk 11?

Comment (cwperks) [reply server/src/main/resources/org/opensearch/bootstrap/security.policy:88]: This is the way that bouncycastle releases artifacts. Until last month the securityplugin was using [jdk15on](https://mvnrepository.com/artifact/org.bouncycastle/bcprov-jdk15on), but then switched to [jdk15to18](https://mvnrepository.com/artifact/org.bouncycastle/bcprov-jdk15to18) (PR where it was changed: https://github.com/opensearch-project/security/pull/2901). I plan to open an issue about using [jdk18on](https://mvnrepository.com/artifact/org.bouncycastle/bcprov-jdk18on).

Comment (reta) [reply server/src/main/resources/org/opensearch/bootstrap/security.policy:88]: No, this is due to https://github.com/opensearch-project/OpenSearch/pull/9289, the BC has moved to core and now we need this permission to manage security providers.

**Comment by willyborankin** [comment]
> Is there a danger that this lowers the security posture of the SSL in such a way that **any** plugins can manipulate the SSL configuration, whereas previously this was isolated to plugins with this policy?

Thats why it is a workaround for 2.10. But it needs to be addressed asap.

**Comment by willyborankin** [comment]
> > > Is there a danger that this lowers the security posture of the SSL in such a way that **any** plugins can manipulate the SSL configuration, whereas previously this was isolated to plugins with this policy?
> > 
> > 
> > Thats why it is a workaround for 2.10. But it needs to be addressed asap.
> 
> @peternied @cwperks @willyborankin the problem is more severe than that, the permissions in this pull request only grant the add/remove provider privileges, but we need much more, the plugin will fail down the line, for example I just added this small piece of code into security plugin initialization sequence:
> 
> ```
>        import org.bouncycastle.jcajce.provider.symmetric.util.GcmSpecUtil;
>  
>        ...
>         if (GcmSpecUtil.gcmSpecExists()) {
>             System.out.println("GcmSpec exists");
>         }
> ```
> 
> It fails with `java.security.AccessControlException: access denied ("java.lang.RuntimePermission" "accessDeclaredMembers")` because, well, `bcprov` needs reflection access as well.

I'm afraid no only the sec plugin. Oh boy oh boy. This is a security hole. Since technically any plugin can chnage SSL behave for example or revoke access to the important functionality.

**Comment by cwperks** [comment]
@willyborankin I see you mentioned there is a performance hit if the security plugin explicitly passes in a provider in calls that take a provider. https://github.com/opensearch-project/security/pull/3317#issuecomment-1707108563

i.e.

```
Cipher cipher = Cipher.getInstance("DESede/CBC/NoPadding", BouncyCastleProvider.PROVIDER_NAME)
```

instead of

```
Cipher cipher = Cipher.getInstance("DESede/CBC/NoPadding");
```

but that there's a performance hit. If the provider is passed in explicitly within the security plugin, which policy file is in effect?

**Comment by willyborankin** [comment]
> @willyborankin I see you mentioned there is a performance hit if the security plugin explicitly passes in a provider in calls that take a provider. [opensearch-project/security#3317 (comment)](https://github.com/opensearch-project/security/pull/3317#issuecomment-1707108563)
> 
> i.e.
> 
> ```
> Cipher cipher = Cipher.getInstance("DESede/CBC/NoPadding", BouncyCastleProvider.PROVIDER_NAME)
> ```
> 
> instead of
> 
> ```
> Cipher cipher = Cipher.getInstance("DESede/CBC/NoPadding");
> ```
> 
> If the provider is passed in explicitly within the security plugin to calls that need one, which policy file is in effect? Any idea what the performance tax is for explicitly passing the provider in each call?

the policy file is OpenSerach requirements.  If library is in `libs` the policy will be shared between all plugins otherwise each plugin is responsible for the policy it needs. We need a better solution in case of crypto-sdk nothing more but I do not know how it should look like.

**Review by peternied — CHANGES_REQUESTED** [review]
From the dialog I see on this issue it looks like this is creating a new category of risk where previously only individual plugins needed to modify their policy file to gain access to restricted APIs, but by adding this privilege up to core there is no longer a clear isolation.

A change to security posture needs to go through a security review process as opposed to an ad-hoc 11th hour patch.

**Comment by cwperks** [comment]
Closing this PR. Any further discussion on this issue should happen in the security issue: https://github.com/opensearch-project/security/issues/3309

**Pull request closed without merging**
